File: src/backend/src/main/java/vn/aims/product/ProductService.java
package vn.aims.product;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Transactional(readOnly = true)
public class ProductService {
    private final ProductRepository repository;
    public ProductService(ProductRepository repository) { this.repository = repository; }

    public List<ProductResponse> search(ProductSearch search) {
        return repository.search(search).stream().map(ProductResponse::from).toList();
    }
    public List<ProductResponse> random() {
        return repository.random().stream().map(ProductResponse::from).toList();
    }
    public ProductDetailResponse detail(int id) {
        Product p = repository.findVisible(id).orElseThrow(() -> new CatalogException(404,"Product with ID " + id + " not found"));
        return switch (p.productType.toUpperCase(Locale.ROOT)) {
            case "BOOK" -> ProductDetailResponse.of(p,"book",book(repository.detail(Book.class,id)));
            case "CD" -> ProductDetailResponse.of(p,"cd",cd(repository.detail(Cd.class,id)));
            case "DVD" -> ProductDetailResponse.of(p,"dvd",dvd(repository.detail(Dvd.class,id)));
            case "NEWSPAPER" -> ProductDetailResponse.of(p,"newspaper",newspaper(repository.detail(Newspaper.class,id)));
            default -> throw new CatalogException(400,"Unsupported productType: " + p.productType);
        };
    }

    // The legacy API flattens media fields and emits all three aliases, even when null.
    private Map<String,Object> media(Media m) {
        Map<String,Object> result = new LinkedHashMap<>();
        result.put("productID",m.productID);
        result.put("publisher",m.publisher);
        result.put("releaseDate",m.releaseDate);
        result.put("language",m.language);
        result.put("genre",m.genre);
        result.put("publicationDate",m.releaseDate);
        result.put("recordLabel",m.publisher);
        result.put("studio",m.publisher);
        return result;
    }
    private Map<String,Object> book(Book b) {
        if (b == null) return null;
        var result=media(b.media);
        result.put("authors",b.authors); result.put("coverType",b.coverType); result.put("numPages",b.numPages);
        return result;
    }
    private Map<String,Object> cd(Cd c) {
        if (c == null) return null;
        var result=media(c.media);
        result.put("artists",c.artists);
        result.put("tracks",c.tracks.stream().map(t -> new TrackResponse(t.id,t.title,t.lengthSeconds)).toList());
        return result;
    }
    private Map<String,Object> dvd(Dvd d) {
        if (d == null) return null;
        var result=media(d.media);
        result.put("discType",d.discType); result.put("director",d.director);
        result.put("runtimeMinutes",d.runtimeMinutes); result.put("subtitles",d.subtitles);
        return result;
    }
    private Map<String,Object> newspaper(Newspaper n) {
        if (n == null) return null;
        var result=media(n.media);
        result.put("editorInChief",n.editorInChief); result.put("issueNumber",n.issueNumber);
        result.put("frequency",n.frequency); result.put("issn",n.issn); result.put("sections",n.sections);
        return result;
    }
    private record TrackResponse(int id, String title, int lengthSeconds) { }
}

File: src/frontend/src/app/services/product.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, Inject } from '@angular/core';
import { Observable } from 'rxjs';
import { API_BASE_URL } from '../app.config';

export interface CdTrack {
  id?: number;
  title: string;
  lengthSeconds: number;
}

export interface Product {
  productID: number;
  productType: 'BOOK' | 'CD' | 'DVD' | 'NEWSPAPER' | string;
  title: string;
  category: string;
  description?: string | null;
  barcode: string;
  length?: number | null;
  width?: number | null;
  height?: number | null;
  weight: number;
  originalPrice: number | string;
  currentPrice: number | string;
  quantityInStock: number;
  status: string;
  imageUrl?: string | null;
  book?: {
    productID: number;
    authors: string;
    coverType: string;
    publisher: string;
    publicationDate: string;
    numPages?: number | null;
    language?: string | null;
    genre?: string | null;
  } | null;
  newspaper?: {
    productID: number;
    editorInChief: string;
    publisher: string;
    publicationDate: string;
    issueNumber?: string | null;
    frequency?: string | null;
    issn?: string | null;
    language?: string | null;
    sections?: string | null;
  } | null;
  cd?: {
    productID: number;
    artists: string;
    recordLabel: string;
    genre: string;
    releaseDate?: string | null;
    tracks?: CdTrack[];
  } | null;
  dvd?: {
    productID: number;
    discType: string;
    director: string;
    runtimeMinutes: number;
    studio: string;
    language: string;
    subtitles: string;
    releaseDate?: string | null;
    genre?: string | null;
  } | null;
}

export interface ProductSearchParams {
  keyword?: string;
  category?: string;
  mediaTypes?: string[];
  minPrice?: number;
  maxPrice?: number;
  status?: string;
}

@Injectable({ providedIn: 'root' })
export class ProductService {
  constructor(
    private http: HttpClient,
    @Inject(API_BASE_URL) private readonly baseUrl: string
  ) {}

  getRandomProducts(): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.baseUrl}/api/products/random`);
  }

  searchProducts(params: ProductSearchParams): Observable<Product[]> {
    let httpParams = new HttpParams();
    if (params.keyword?.trim()) {
      httpParams = httpParams.set('keyword', params.keyword.trim());
    }
    if (params.category) {
      httpParams = httpParams.set('category', params.category);
    }
    if (params.mediaTypes?.length) {
      httpParams = httpParams.set('mediaTypes', params.mediaTypes.join(','));
    }
    if (params.minPrice !== undefined) {
      httpParams = httpParams.set('minPrice', params.minPrice);
    }
    if (params.maxPrice !== undefined) {
      httpParams = httpParams.set('maxPrice', params.maxPrice);
    }
    if (params.status) {
      httpParams = httpParams.set('status', params.status);
    }

    return this.http.get<Product[]>(`${this.baseUrl}/api/products`, { params: httpParams });
  }

  getProductById(id: number): Observable<Product> {
    return this.http.get<Product>(`${this.baseUrl}/api/products/${id}`);
  }

  createProduct(dto: any): Observable<Product> {
    return this.http.post<Product>(`${this.baseUrl}/api/products`, dto);
  }

  updateProduct(id: number, dto: any): Observable<Product> {
    return this.http.patch<Product>(`${this.baseUrl}/api/products/${id}`, dto);
  }

  adjustStock(id: number, quantityDelta: number, reason: string): Observable<Product> {
    return this.http.patch<Product>(`${this.baseUrl}/api/products/${id}/stock`, {
      quantityDelta,
      reason
    });
  }

  deleteProducts(ids: number[]): Observable<any> {
    return this.http.post<any>(`${this.baseUrl}/api/products/batch-delete`, { ids });
  }

  deactivateProducts(ids: number[]): Observable<any> {
    return this.http.post<any>(`${this.baseUrl}/api/products/batch-deactivate`, { ids });
  }

  getAuditLogs(): Observable<any[]> {
    return this.http.get<any[]>(`${this.baseUrl}/api/products/audit-logs`);
  }
}
